### PR TITLE
BUG: interpolate.LinearNDInterpolator: fix for precomputed triangulation

### LINKED ISSUE #21043: BUG: scipy.interpolate.LinearNDInterpolator call _calculate_triangulation with Delaunay triangulation
### Describe your issue.

There was a re-factor on "interpnd.py" that add a "_calculate_triangulation" function:
- https://github.com/scipy/scipy/blob/main/scipy/interpolate/interpnd.pyx#L304
it is always call from here: 
- https://github.com/scipy/scipy/blob/main/scipy/interpolate/interpnd.pyx#L92

So, now, if LinearNDInterpolator is built with a precomputed Delaunay triangulation, a triangulation is recalculated as if only points had been passed.

This breaks all the code that gives its own triangulation.

### Possible Code Correction

```python
# old code was (V1.11.x):
class LinearNDInterpolator(NDInterpolatorBase):
    def __init__(self, points, values, fill_value=np.nan, rescale=False):
        NDInterpolatorBase.__init__(self, points, values, fill_value=fill_value, rescale=rescale)
        if self.tri is None:
            self.tri = qhull.Delaunay(self.points)


# may be corrected by something like that on NDInterpolatorBase constructor
if (self.tri = None) :  # add control before call _calculate_triangulation.
    self._calculate_triangulation(self.points)
```

### Error message

```shell
No error report, but wrong result.
```

### SciPy/NumPy/Python version and system information

```shell
SciPy 1.12/1.13/1.13.1
```

### PR DESCRIPTION
**Pull request by Djip007**
#### Reference issue
https://github.com/scipy/scipy/issues/21043

#### Test
```
python dev.py test -s interpolate -m full
[...]
scipy/interpolate/tests/test_interpnd.py ......................                                                                                        [ 33%]
[...]
======================================================== 1023 passed, 17 skipped, 1 xfailed in 11.38s ========================================================
```
look no break on existing test.

### CODE REVIEW

**Review by j-bowhay — COMMENTED** [review]
Is it possible to add a regression test to ensure that if a triangulation is passed that triangulation is retained?

**Comment by Djip007** [comment]
> Is it possible to add a regression test to ensure that if a triangulation is passed that triangulation is retained?

I will try...
I find where I need to write the test.
Need to figure how to do a simple test for this use case... may need some time.

**Comment by Djip007** [comment]
a simple test can be to control that the triangulation passed is not modified. With simple add in this test
```python
# Actuel
    def test_tri_input(self):
        # Test at single points
        x = np.array([(0,0), (-0.5,-0.5), (-0.5,0.5), (0.5, 0.5), (0.25, 0.3)],
                     dtype=np.float64)
        y = np.arange(x.shape[0], dtype=np.float64)
        y = y - 3j*y

        tri = qhull.Delaunay(x)
        yi = interpnd.LinearNDInterpolator(tri, y)(x)

        assert_almost_equal(y, yi)

# Modified
    def test_tri_input(self):
        # Test at single points
        x = np.array([(0,0), (-0.5,-0.5), (-0.5,0.5), (0.5, 0.5), (0.25, 0.3)],
                     dtype=np.float64)
        y = np.arange(x.shape[0], dtype=np.float64)
        y = y - 3j*y

        tri = qhull.Delaunay(x)
        interpolator = interpnd.LinearNDInterpolator(tri, y)
        yi = interpolator(x)

        assert_almost_equal(y, yi)
        # insure the triangulation is not recompute.
        assert interpolator.tri is tri
```
To be fair, I don't rely like that :wink:

**Comment by ev-br** [comment]
Don't change existing tests, just add a new one

**Comment by Djip007** [comment]
> Don't change existing tests, just add a new one

I see... in the same vein: is it best to make 2 tests than have 2 check in 1 test ?

**Comment by ev-br** [comment]
Adding an assert to an existing test is fine. 
Just send a PR and the diff will be clear if it's ok or not.

**Comment by Djip007** [comment]
> add a regression test to ensure that if a triangulation is passed that triangulation is retained

Add 1 test for it. it pass with this patch, failed without for me.

:crossed_fingers:

**Comment by Djip007** [comment]
correct linting ... I hope.

File: scipy/interpolate/tests/test_interpnd.py [hunk 1]
```diff
@@ -62,6 +62,41 @@ def test_tri_input(self):
...
+                                           (-1,-1,-1),
+                                           (-1,-1,-1)], dtype=np.int32)
+                self.ndim = self._points.shape[1]
+                self.npoints = self._points.shape[0]
+                self.min_bound = self._points.min(axis=0)
+                self.max_bound = self._points.max(axis=0)
+                self.equations = np.zeros((9, 4), dtype=np.double)
+                self.paraboloid_scale = 1
+                self.paraboloid_shift = 0
+                self._transform = None
```
Comment (j-bowhay) [inline scipy/interpolate/tests/test_interpnd.py:87]: Is there a reason for not just using `Delaunay` directly, this seems potentially fragile if the internal implementation of `Delaunay` changes?

Comment (Djip007) [reply scipy/interpolate/tests/test_interpnd.py:87]: Yes I have a reason, but may be not a good one.
There is already a test that create a Delaunay triangulation.
In that case I want to be sure that this triangulation is not recompute. So I try to give the interpolation a "FakeDelaunay" that do not have same result if it is recompute. 

But I can perfectly understand your point. 
Tell me what you think. I can only keep the `assert interpolator.tri is tri` and possibly "merge" this assert on the `test_tri2_input` test.

**Comment by ev-br** [comment]
> I can only keep the assert interpolator.tri is tri and possibly "merge" this assert on the test_tri2_input test.

Yes please (disregard my comment above, the diff is clear)

**Review by j-bowhay — APPROVED** [review]
Thanks @Djip007 and congratulations on your first contribution to SciPy!

**Comment by Djip007** [comment]
Thank you for your help.

**Pull request merged**